hax/hax/hax.py

```diff
@@ -56,6 +57,30 @@ def _run_stats_updater_thread(motr: Motr,
     return thread
 
 
+@repeat_if_fails()
+def _remove_stale_session(util: ConsulUtil) -> None:
+    """
+    Destroys a stale RC leader session if it exists or does nothing otherwise.
+
+    An RC leader session may survive 'hctl shutdown'. In such a case 'leader'
+    key will contain a garbage value 'elect2805' but the session will be alive
+    and thus RC leader will not be re-elected.
+    """
+    if not util.kv.kv_get_raw('leader'):
+        # No leader key means that there can be no stale RC leader session for
+        # sure. We're starting for the first time against a fresh Consul KV.
+        return
+
+    sess = util.get_leader_session_no_wait()
+    node = util.get_leader_node()
+    if re.match(r'^elect[\d]+$', node):
```

Comment:
If some node hostname will be `elect123` or something - it would match this regexp and would cause session destructions on each `hax` startup on this node. Why not check explicitly on the magic `elect2805` value?

Reply:
Because any `^elect[\d]+$` is considered to be a stub. 

https://github.com/Seagate/cortx-hare/blob/ce3411e975b4cb93f2e456e1c5f3832191535844/utils/hare-bootstrap#L112-L123

In other words, whole bootstrapping and RC re-election will probably be broken if some of the hostnames match that regexp :-) And this doesn't seem to be the problem of this PR.

Reply:
No, it won't be broken in case of bootstrapping via hctl, because we are checking for the session, not the hostname value there.

Reply:
Anyway we have other parts of the code that do have assumption that `elect[\d]+` means "No leader elected". If we don't have that assumption anymore, then other parts of the code should be adjusted.

Reply:
Probably. I'm not aware of the other parts, but I don't think that it's right to use it here as a pattern instead of a concrete value.

Reply:
`elect2805` gets written from hare-shutdown by your suggestion (we need to put "fake" leader node there to enable RC re-election to happen when the hare cluster is restarted again). `elect2805` is chosen as a particular value that corresponds to the code below:
https://github.com/Seagate/cortx-hare/blob/ce3411e975b4cb93f2e456e1c5f3832191535844/utils/hare-bootstrap#L112-L123
(I guess you are the author)

Regarding other parts. This is another example where this pattern is already used:
https://github.com/Seagate/cortx-hare/blob/ce3411e975b4cb93f2e456e1c5f3832191535844/utils/utils.py#L167-L168

BTW this function appeared quite long ago: https://github.com/Seagate/cortx-hare/pull/728

Long story short, 
1. there is (perhaps implicit) assumption that sometimes `leader` contains the text which is NOT a hostname. It obviously means that not all hostnames are suitable for Hare cluster, but that decision seems to have been made at the very beginning of Hare history.
2. In this case the value is described by `^elect[\d]+$` regexp.
3. Because of \[1] some values will always be considered as "stub" instead of "real hostname", whatever regexp to describe it we use.

That's why I suggest to move this discussion out of this PR. If you think there is a way to consider that any value in `leader` always means hostname, let's create a ticket and address it separately.

Reply:
Let me try to explain it again: the code at `hare-bootstrap` does not assume "that sometimes leader contains the text which is NOT a hostname". This code just writes some value to leader in order to trigger the RC leader election mechanism, but it does not check the value. So again - it does not assume anything about it.

The problem is with the code, which indeed makes such assumptions - I have never written it and was not aware about it. In any case, we better don't extend such assumptions in our code - it's wrong and dangerous.

Reply:
I am okay with having a specific magic value to represent an invalid session. That sounds reasonable.
I think `consul kv put leader elect$RANDOM`  triggered the pattern matching which is also done in some other places.
Now, to trigger RC leader during `hctl bootstrap` we need to change the leader value and having 1 magic value will not work as it will not trigger the change in the leader KV. I think we can keep `consul kv put leader elect$RANDOM` as its the bootstrap path, in case of failure to bootstrap the cluster won't be usable anyway.
The problem mainly lies when the services are stopped and started outside `hctl bootstrap`, mainly through conrtx-ha (via pacemaker). Presently as part of hctl shutdown we are resetting the leader KV, here we can use a defined magic for resetting.
@knekrasov, I wonder if we have really solved the problem as we are destroying the session but will it really re-trigger RC leader election in case of hax stop - start sequences because we are resetting the leader KV in `hare-shutdown` but not during hax shutdown.